### PR TITLE
Milestone Implementation of API Composition feature

### PR DESCRIPTION
**Pull request by praminda**
### Proposed changes in this pull request
- Stable milestone  implementation of Composite APIs feature

### When should this PR be merged
After PR Review

#### General

- [ ] **Is this PR explained thoroughly?** All code changes must be accounted for in the PR description.
- [ ] **Is the PR labeled correctly?**

#### Functionality

- [ ] **Are all requirements met?** Compare implemented functionality with the requirements specification.
- [ ] **Does the UI work as expected?** There should be no Javascript errors in the console; all resources should load. There should be no unexpected errors. Deliberately try to break the feature to find out if there are corner cases that are not handled.

#### Code

- [ ] **Do you fully understand the introduced changes to the code?** If not ask for clarification, it might uncover ways to solve a problem in a more elegant and efficient way.
- [ ] **Does the PR introduce any inefficient database requests?** Use the debug server to check for duplicate requests.
- [ ] **Are all necessary strings marked for translation?** All strings that are exposed to users via the UI must be [marked for translation](https://docs.djangoproject.com/en/1.10/topics/i18n/translation/).

#### Tests

- [ ] **Are there sufficient test cases?** Ensure that all components are tested individually; models, forms, and serializers should be tested in isolation even if a test for a view covers these components.
- [ ] **If this is a bug fix, are tests for the issue in place?**  There must be a test case for the bug to ensure the issue won’t regress. Make sure that the tests break without the new code to fix the issue.
- [ ] **If this is a new feature or a significant change to an existing feature?** has the manual testing spreadsheet been updated with instructions for manual testing?

#### Security

- [ ] **Confirm this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets.**
- [ ] **Are all UI and API inputs run through forms or serializers?**
- [ ] **Are all external inputs validated and sanitized appropriately?**
- [ ] **Does all branching logic have a default case?**
- [ ] **Does this solution handle outliers and edge cases gracefully?**
- [ ] **Are all external communications secured and restricted to SSL?**

#### Documentation

- [ ] **Are changes to the UI documented in the platform docs?** If this PR introduces new platform site functionality or changes existing ones, the changes should be documented.
- [ ] **Are changes to the API documented in the API docs?** If this PR introduces new API functionality or changes existing ones, the changes must be documented.
- [ ] **Are reusable components documented?** If this PR introduces components that are relevant to other developers (for instance a mixin for a view or a generic form) they should be documented in the Wiki.

### CODE REVIEW

File: components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/util/APIUtils.java [hunk 1]
```diff
@@ -193,6 +220,27 @@ public static String generateOperationIdFromPath(String path, String httpVerb) {
...
+    /**
+     * Generates default URI templates to be assigned to an API
+     * @return  Map of URI Templates
+     */
+    public static Map<String, UriTemplate> getDefaultUriTemplates() {
+        Map<String, UriTemplate> uriTemplateMap = new HashMap<>();
+        UriTemplate.UriTemplateBuilder uriTemplateBuilder = new UriTemplate.UriTemplateBuilder();
+        uriTemplateBuilder.uriTemplate("/*");
+        for (String httpVerb : APIMgtConstants.SUPPORTED_HTTP_VERBS.split(",")) {
+            if (!"OPTIONS".equals(httpVerb)) {
```
Comment (bhathiya) [inline components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/util/APIUtils.java]: Let's use a constant here.

File: components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/dao/ApiDAO.java [hunk 2]
```diff
@@ -59,11 +63,50 @@
...
+     * Retrieves a paginated list of composite APIs. Resulting APIs will be filtered for
+     * PROVIDER name <code>user</code>.
+     *
+     * @param roles     If provided, results will be filtered for only the APIs with these roles.
+     *                  If not APIs with any role will be retrieved.
+     * @param user      Name of the PROVIDER who owns requested set of Composite APIs
+     * @param offset    Page number
+     *                  <p>(ex: <code>offset</code> 2 with <code>limit</code> 10 will retrieve 11-20th results)</p>
+     * @param limit     Number of results to be included in the resulting list
+     * @return
```
Comment (praminda) [inline components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/dao/ApiDAO.java]: Fixed with  0822410

File: components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/api/APIDefinition.java [hunk 3]
```diff
@@ -59,6 +60,13 @@
     String generateSwaggerFromResources(API.APIBuilder api);
 
     /**
+     * generate the swagger from uri templates.
```
Comment (praminda) [inline components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/api/APIDefinition.java]: Fixed with  0822410

File: components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/api/APIMgtAdminService.java [hunk 4]
```diff
@@ -165,15 +165,17 @@
 
     /**
      * Get a list of APIs with given gateway labels and status
```
Comment (praminda) [inline components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/api/APIMgtAdminService.java]: Fixed with  0822410

File: components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/dao/impl/ApiDAOImpl.java [hunk 5]
```diff
@@ -135,6 +154,46 @@ public API getAPISummary(String apiID) throws APIMgtDAOException {
...
+                PreparedStatement statement = connection.prepareStatement(query)) {
+            statement.setString(1, apiID);
+            statement.setString(2, ApiType.COMPOSITE.toString());
+
+            List<CompositeAPI> apiResults = getCompositeAPISummaryList(connection, statement);
+            if (apiResults.isEmpty()) {
+                return null;
+            }
+
+            // there should be on 1 result from the database
```
Comment (praminda) [inline components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/dao/impl/ApiDAOImpl.java]: Fixed with  0822410

File: components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/dao/impl/ApiDAOImpl.java [hunk 6]
```diff
@@ -505,74 +549,45 @@ public boolean isAPIContextExists(String contextName) throws APIMgtDAOException
...
+            throw new APIMgtDAOException(e);
+        }
+    }
 
-                if (API.Visibility.RESTRICTED == api.getVisibility()) {
-                    addVisibleRole(connection, apiPrimaryKey, api.getVisibleRoles());
-                }
+    /**
+     * {@inheritDoc}
+     */
```
Comment (praminda) [inline components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/dao/impl/ApiDAOImpl.java]: Fixed with  0822410

File: components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/dao/impl/ApiDAOImpl.java [hunk 7]
```diff
@@ -586,6 +601,108 @@ public void addAPI(final API api) throws APIMgtDAOException {
     }
 
     /**
+     * Method for adding API related information
```
Comment (praminda) [inline components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/dao/impl/ApiDAOImpl.java]: Fixed with  0822410

File: components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/dao/impl/ApiDAOImpl.java [hunk 8]
```diff
@@ -586,6 +601,108 @@ public void addAPI(final API api) throws APIMgtDAOException {
...
+        addUrlMappings(connection, api.getUriTemplates().values(), apiPrimaryKey);
+        addSubscriptionPolicies(connection, api.getPolicies(), apiPrimaryKey);
+        addEndPointsForApi(connection, apiPrimaryKey, api.getEndpoint());
+        addAPIDefinition(connection, apiPrimaryKey, api.getApiDefinition(), api.getCreatedBy());
+        addAPIPermission(connection, api.getPermissionMap(), apiPrimaryKey);
+    }
+
+
+    /**
+     * Method for adding Composite API related information
```
Comment (praminda) [inline components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/dao/impl/ApiDAOImpl.java]: Fixed with  0822410

File: components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/dao/impl/ApiDAOImpl.java [hunk 9]
```diff
@@ -689,39 +806,73 @@ public void updateAPI(String apiID, API substituteAPI) throws APIMgtDAOException
...
+             PreparedStatement statement = connection.prepareStatement(API_DELETE)) {
+            persistAPIDelete(connection, statement, apiID);
+        } catch (SQLException | IOException e) {
             throw new APIMgtDAOException(e);
         }
     }
 
     /**
+     * {@inheritDoc}
+     */
```
Comment (praminda) [inline components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/dao/impl/ApiDAOImpl.java]: Fixed with  0822410

File: components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/dao/impl/ApiDAOImpl.java [hunk 10]
```diff
@@ -689,39 +806,73 @@ public void updateAPI(String apiID, API substituteAPI) throws APIMgtDAOException
...
+    @Override
+    public void deleteCompositeApi(String apiId) throws APIMgtDAOException {
+        APISubscriptionDAO apiSubscriptionDAO = DAOFactory.getAPISubscriptionDAO();
+        List<Subscription> subscriptions = apiSubscriptionDAO.getAPISubscriptionsByAPI(apiId);
+
+        for (Subscription subscription : subscriptions) {
+            apiSubscriptionDAO.deleteAPISubscription(subscription.getId());
+        }
+
+
```
Comment (praminda) [inline components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/dao/impl/ApiDAOImpl.java]: Fixed with  0822410

File: components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/dao/impl/ApiDAOImpl.java [hunk 11]
```diff
@@ -730,15 +881,30 @@ public String getSwaggerDefinition(String apiID) throws APIMgtDAOException {
     }
 
     /**
-     * @see ApiDAO#updateSwaggerDefinition(String, String, String)
+     * {@inheritDoc}
+     */
```
Comment (praminda) [inline components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/dao/impl/ApiDAOImpl.java]: Fixed with  0822410

File: components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/dao/impl/ApiDAOImpl.java [hunk 12]
```diff
@@ -730,15 +881,30 @@ public String getSwaggerDefinition(String apiID) throws APIMgtDAOException {
...
+
+            return IOUtils.toString(apiDefinition, StandardCharsets.UTF_8);
+        } catch (SQLException | IOException e) {
+            throw new APIMgtDAOException(e);
+        }
+    }
+
+    /**
+     * @see ApiDAO#updateApiDefinition(String, String, String)
      */
```
Comment (praminda) [inline components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/dao/impl/ApiDAOImpl.java]: Fixed with  0822410

File: components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/impl/APIDefinitionFromSwagger20.java [hunk 13]
```diff
@@ -249,23 +287,41 @@ public String generateSwaggerFromResources(API.APIBuilder api)  {
...
+        SwaggerParser swaggerParser = new SwaggerParser();
+        Swagger swagger = swaggerParser.parse(apiDefinition);
+
+        if (swagger == null) {
+            throw new APIManagementException("Swagger could not be generated from provided API definition");
+        }
+
+        Info apiInfo = swagger.getInfo();
+        if (apiInfo == null) {
+            throw new APIManagementException("Swagger doesn't contains the info");
```
Comment (praminda) [inline components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/impl/APIDefinitionFromSwagger20.java]: Fixed with  0822410

File: components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/impl/APIGatewayPublisherImpl.java [hunk 14]
```diff
@@ -137,6 +163,21 @@ public void deleteAPI(API api) throws GatewayException {
         }
```
Comment (praminda) [inline components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/impl/APIGatewayPublisherImpl.java]: Fixed with  0822410

File: components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/impl/APIGatewayPublisherImpl.java [hunk 15]
```diff
@@ -137,6 +163,21 @@ public void deleteAPI(API api) throws GatewayException {
...
+            // build the message to send
+            APIDTO gatewayDTO = new APIDTO(APIMgtConstants.GatewayEventTypes.API_DELETE);
+            gatewayDTO.setLabels(api.getLabels());
+            APISummary apiSummary = new APISummary(api.getId());
+            apiSummary.setName(api.getName());
+            apiSummary.setVersion(api.getVersion());
+            apiSummary.setContext(api.getContext());
+            gatewayDTO.setApiSummary(apiSummary);
+            publishToPublisherTopic(gatewayDTO);
+        }
```
Comment (praminda) [inline components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/impl/APIGatewayPublisherImpl.java]: Fixed with  0822410

File: components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/impl/APIGatewayPublisherImpl.java [hunk 16]
```diff
@@ -228,11 +269,12 @@ private void publishToStoreTopic(GatewayDTO gatewayDTO) throws GatewayException
     /**
      * Save API into FS
      *
-     * @param api     API object
+     * @param apiName     API Name
+     * @param apiVersion     API Version
      * @param gwHome  path of the gateway
      * @param content API config
```
Comment (praminda) [inline components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/impl/APIGatewayPublisherImpl.java]: Fixed with  0822410

File: components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/impl/APIStoreImpl.java [hunk 17]
```diff
@@ -889,37 +943,28 @@ private void setGatewayDefinitionSource(API.APIBuilder apiBuilder) throws APITem
             resourceList.add(dto);
         }
         GatewaySourceGenerator gatewaySourceGenerator = getGatewaySourceGenerator();
-        gatewaySourceGenerator.setAPI(apiBuilder.build());
-        String gatewayConfig = gatewaySourceGenerator.getConfigStringFromTemplate(resourceList);
+        APIConfigContext apiConfigContext = new APIConfigContext(apiBuilder.getName(), apiBuilder.getContext(),
+                apiBuilder.getVersion(), apiBuilder.getCreatedTime(), config.getGatewayPackageName());
+
+
```
Comment (praminda) [inline components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/impl/APIStoreImpl.java]: Fixed with  0822410

File: components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/impl/APIStoreImpl.java [hunk 18]
```diff
@@ -1156,6 +1205,76 @@ public String addCompositeApiFromDefinition(String swaggerResourceUrl) throws AP
         return apiResults;
     }
 
+    @Override
+    public List<CompositeAPI> searchCompositeAPIs(String query, int offset, int limit) throws APIManagementException {
+        List<CompositeAPI> apiResults;
+
+        //this should be current logged in user
+        String user = "admin";
```
Comment (praminda) [inline components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/impl/APIStoreImpl.java]: Fixed with  0822410

File: components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/util/APIUtils.java (hunk 1 above)
Comment (praminda) [inline components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/util/APIUtils.java]: Fixed with  0822410

**Pull request merged**